File: agent_go/cmd/family-server/voice_worker.go
package main

import (
	"bufio"
	"context"
	"embed"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"sync"
	"time"
)

//go:embed voice_worker.py
var voiceWorkerScript embed.FS

// voiceWorkerIdleTimeout is how long the warm worker stays loaded after its
// last request before shutting itself down. Mirrors this app's own
// download-on-activate / delete-on-deactivate philosophy, applied to RUNTIME
// memory instead of disk: a family mid-conversation keeps the model warm
// (each call ~0.8s instead of ~3.5s), but an idle process isn't left holding
// onto RAM/GPU memory for a family who stepped away.
const voiceWorkerIdleTimeout = 15 * time.Minute

// voiceWorker manages ONE persistent Python process that keeps the Parakeet
// (STT) model loaded in memory, reused across every request rather than
// reloaded from scratch each time. A fresh
// `python -c` process pays ~1.9s just re-importing mlx_audio before any real
// work happens — measured directly, not assumed — which is why this exists.
//
// Requests are strictly serialized (one in flight at a time): MLX/Metal has
// documented issues under concurrent access (mlx-audio's own GitHub reports
// crashes under concurrent requests), and a family's voice usage is
// inherently serial anyway — one recording or one reply spoken at a time.
type voiceWorker struct {
	mu     sync.Mutex
	cmd    *exec.Cmd
	stdin  io.WriteCloser
	reader *bufio.Reader
	idle   *time.Timer
	// warm tracks which models are ACTUALLY loaded in the running process's
	// memory right now — distinct from "installed on disk". A model can be
	// fully installed but still cold (worker never started, or idle-shut-down
	// 15 minutes ago) — this is what lets the UI say "Warming up..." instead
	// of falsely claiming instant readiness.
	warm map[string]bool

	// How to start the process. Everything else here — the readiness
	// handshake, call timeout, idle unload, crash reaping — is transport
	// agnostic, so the native Swift helper (voice_native.go) reuses all of it
	// by supplying a different launcher rather than duplicating the
	// supervision logic. name only labels log lines.
	name   string
	launch func() (*exec.Cmd, error)
	// Per-worker override for voiceWorkerCallTimeout. The native helper needs
	// a much longer one: its very first `load` downloads CoreML weights
	// (~96s measured), which would otherwise trip the 30s default and get the
	// process killed mid-download, forever.
	callTimeout time.Duration
	// Per-worker override for voiceWorkerIdleTimeout.
	idleTimeout time.Duration
}

func (w *voiceWorker) idleAfter() time.Duration {
	if w.idleTimeout > 0 {
		return w.idleTimeout
	}
	return voiceWorkerIdleTimeout
}

func (w *voiceWorker) timeout() time.Duration {
	if w.callTimeout > 0 {
		return w.callTimeout
	}
	return voiceWorkerCallTimeout
}

var sharedVoiceWorker = &voiceWorker{name: "voice", launch: pythonVoiceWorkerCmd}

// pythonVoiceWorkerCmd starts the MLX/Python worker.
func pythonVoiceWorkerCmd() (*exec.Cmd, error) {
	scriptBytes, err := voiceWorkerScript.ReadFile("voice_worker.py")
	if err != nil {
		return nil, err
	}
	// Written into the SAME directory the shared MLX environment already
	// lives in, rather than embedded-and-run-from-memory, so it's a normal
	// file `python` can execute and a developer can open to debug.
	scriptPath := filepath.Join(mlxVoiceDir(), "voice_worker.py")
	if err := os.WriteFile(scriptPath, scriptBytes, 0o600); err != nil {
		return nil, err
	}
	return exec.Command(mlxVoicePython(), scriptPath), nil
}

// voiceWorkerCallTimeout bounds how long ANY single request may block waiting
// on the worker's response. Without this, a genuinely stuck worker (observed
// live: the Python process sitting at ~0% CPU, never responding, after a
// transcribe request) held w.mu FOREVER — every later call, even a totally
// unrelated one from a different conversation minutes later, piled up behind
// the same dead read with no error, no log line, nothing. Generous enough for
// a real cold start (measured worst case ~7s) plus margin, short enough that
// a genuinely wedged worker gets NOTICED and replaced within one request
// instead of taking the whole voice feature down silently for good.
const voiceWorkerCallTimeout = 30 * time.Second

// call sends one request and waits for its matching response, bounded by
// ctx AND voiceWorkerCallTimeout (whichever is shorter). Starts the worker if
// it isn't already running (first use, previously idle-shut-down, or
// crashed), and restarts+retries ONCE transparently if the pipe is
// unexpectedly broken — a stale/crashed process should be invisible to the
// caller, not a hard failure on their very next click. A timeout counts as
// broken too: the worker is killed so its stuck goroutine unblocks and the
// NEXT call (this retry, or a future one) starts a fresh process rather than
// queuing behind the same dead one forever.
func (w *voiceWorker) call(ctx context.Context, req map[string]any) (map[string]any, error) {
	w.mu.Lock()
	defer w.mu.Unlock()

	callCtx, cancel := context.WithTimeout(ctx, w.timeout())
	defer cancel()
	resp, err := w.callLocked(callCtx, req)
	if err != nil {
		// One retry against a freshly (re)started process — covers the
		// common case of the worker having crashed, been idle-killed, or
		// timed out between calls, without making every transient hiccup
		// visible.
		w.stopLocked()
		retryCtx, retryCancel := context.WithTimeout(ctx, w.timeout())
		defer retryCancel()
		resp, err = w.callLocked(retryCtx, req)
	}
	return resp, err
}

func (w *voiceWorker) callLocked(ctx context.Context, req map[string]any) (map[string]any, error) {
	if err := w.ensureStartedLocked(); err != nil {
		return nil, err
	}
	line, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}
	if _, err := w.stdin.Write(append(line, '\n')); err != nil {
		return nil, fmt.Errorf("voice worker pipe broken: %w", err)
	}
	// ReadString blocks on the pipe with no deadline of its own — run it in a
	// goroutine and race it against ctx so a wedged worker can't hold this
	// call (and w.mu, and every call after it) forever. The goroutine itself
	// may still be blocked in ReadString after we give up on it; killing the
	// process below (via the caller's stopLocked on error) is what actually
	// unblocks it, not this select.
	type readResult struct {
		line string
		err  error
	}
	resultCh := make(chan readResult, 1)
	go func() {
		respLine, err := w.reader.ReadString('\n')
		resultCh <- readResult{respLine, err}
	}()
	var respLine string
	select {
	case res := <-resultCh:
		if res.err != nil {
			return nil, fmt.Errorf("voice worker did not respond: %w", res.err)
		}
		respLine = res.line
	case <-ctx.Done():
		return nil, fmt.Errorf("voice worker timed out: %w", ctx.Err())
	}
	var resp map[string]any
	if err := json.Unmarshal([]byte(respLine), &resp); err != nil {
		return nil, fmt.Errorf("voice worker sent an unreadable response: %w", err)
	}
	if errMsg, ok := resp["error"]; ok {
		return nil, fmt.Errorf("%v", errMsg)
	}

	// Every successful call pushes the idle-shutdown clock back out — the
	// worker only unloads after a real gap in use, not a fixed wall-clock
	// window from when it started.
	if w.idle != nil {
		w.idle.Reset(w.idleAfter())
	}
	if model, ok := req["model"].(string); ok {
		if w.warm == nil {
			w.warm = map[string]bool{}
		}
		w.warm[model] = true
	}
	return resp, nil
}

// IsWarm reports whether a model is ACTUALLY loaded in the worker's memory
// right now — not just installed on disk. False whenever the worker isn't
// running at all (never started, idle-shut-down, or crashed) or hasn't yet
// been asked to use this specific model since it last started.
func (w *voiceWorker) IsWarm(model string) bool {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.cmd != nil && w.warm[model]
}

func (w *voiceWorker) ensureStartedLocked() error {
	if w.cmd != nil {
		return nil
	}
	cmd, err := w.launch()
	if err != nil {
		return err
	}
	stdin, err := cmd.StdinPipe()
	if err != nil {
		return err
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return err
	}
	stderr, err := cmd.StderrPipe()
	if err != nil {
		return err
	}
	if err := cmd.Start(); err != nil {
		return fmt.Errorf("could not start the voice engine: %w", err)
	}

	// Wait for the explicit readiness line on stderr rather than a fixed
	// sleep — import time varies by machine, and stdout is reserved for JSON
	// responses only, so readiness can't be signaled there.
	ready := make(chan error, 1)
	go func() {
		sc := bufio.NewScanner(stderr)
		for sc.Scan() {
			if sc.Text() == "WORKER_READY" {
				ready <- nil
				go io.Copy(io.Discard, stderr) // drain remaining stderr so the pipe never fills and blocks the process
				return
			}
		}
		ready <- fmt.Errorf("voice engine exited before it was ready")
	}()
	select {
	case err := <-ready:
		if err != nil {
			_ = cmd.Process.Kill()
			return err
		}
	case <-time.After(2 * time.Minute):
		_ = cmd.Process.Kill()
		return fmt.Errorf("voice engine took too long to start")
	}

	w.cmd = cmd
	w.stdin = stdin
	w.reader = bufio.NewReader(stdout)
	w.idle = time.AfterFunc(w.idleAfter(), func() {
		w.mu.Lock()
		defer w.mu.Unlock()
		log.Printf("[%s] worker idle for %s, unloading", w.name, w.idleAfter())
		w.stopLocked()
	})

	// Clear state the MOMENT the process actually exits — crash, killed
	// externally, anything — rather than leaving IsWarm()/callLocked() to
	// find out lazily on the next attempted use. Without this, a dead
	// process could be reported "warm" for an arbitrarily long stretch:
	// stale bookkeeping, not the process's real, current state.
	deadCmd := cmd
	go func() {
		// The ONE place that calls Wait() on this process — see stopLocked's
		// doc comment for why that has to be exclusive.
		_ = deadCmd.Wait()
		w.mu.Lock()
		defer w.mu.Unlock()
		if w.cmd == deadCmd { // only clear if nothing else already replaced it
			log.Printf("[voice] worker process exited")
			w.stopLocked()
		}
	}()

	log.Printf("[%s] worker started (models load lazily on first use of each)", w.name)
	return nil
}

// stopLocked asks a still-running process to terminate and clears bookkeeping
// immediately, but does NOT call Wait() itself — the reaper goroutine spawned
// in ensureStartedLocked is already watching this exact process and will Wait
// on it once it actually exits. Calling Wait() from two places on the same
// *exec.Cmd is not safe in Go (a second call errors with "Wait was already
// called"), so exactly one place — the reaper — owns that. Safe to call when
// nothing is running.
func (w *voiceWorker) stopLocked() {
	if w.idle != nil {
		w.idle.Stop()
		w.idle = nil
	}
	if w.cmd != nil && w.cmd.Process != nil {
		_ = w.cmd.Process.Kill()
	}
	w.cmd = nil
	w.stdin = nil
	w.reader = nil
	w.warm = nil // the new process starts cold — nothing is loaded yet
}

File: agent_go/cmd/family-server/voice_worker.py
import sys
import json
import os
import contextlib

# Persistent MLX voice worker. Started once by voice_worker.go and kept warm
# for as long as the family is actively using voice features — this exists
# because a fresh `python -c` process pays ~1.9s just re-importing mlx_audio
# EVERY call, before any real work happens (measured directly: ~1.9s import,
# ~0.8s actual model-load + inference). Loading each model ONCE here and
# reusing the in-memory object for every subsequent request eliminates both
# the repeated import cost and the repeated model reload.

_stt_models = {}


def get_stt_model(name):
    if name not in _stt_models:
        from mlx_audio.stt.utils import load_model
        _stt_models[name] = load_model(name)
    return _stt_models[name]


def _silent_warmup_wav(path, seconds=1.0, sample_rate=16000):
    """A trivial, silent WAV file — good enough to trigger MLX's real
    first-inference compilation (which cares about input SHAPE, not content)."""
    import wave
    import struct
    n_samples = int(seconds * sample_rate)
    with wave.open(path, "wb") as w:
        w.setnchannels(1)
        w.setsampwidth(2)
        w.setframerate(sample_rate)
        w.writeframes(struct.pack(f"<{n_samples}h", *([0] * n_samples)))


def handle_load_stt(req):
    # Loading weights is NOT enough to make the first real use fast: MLX
    # lazily compiles/JIT's the computation graph on the first actual
    # inference, separately from loading weights. A real (discarded)
    # transcription here pays that one-time cost during install/startup
    # instead of on the parent's first real use.
    from mlx_audio.stt.generate import generate_transcription
    import tempfile
    model = get_stt_model(req["model"])
    with tempfile.TemporaryDirectory() as tmp:
        clip = os.path.join(tmp, "warmup.wav")
        _silent_warmup_wav(clip)
        generate_transcription(model=model, audio=clip, verbose=False)
    return {"status": "ok"}


def handle_transcribe(req):
    from mlx_audio.stt.generate import generate_transcription
    model = get_stt_model(req["model"])
    result = generate_transcription(model=model, audio=req["audio_path"], verbose=False)
    return {"text": result.text}


HANDLERS = {
    "transcribe": handle_transcribe,
    "load_stt": handle_load_stt,
}


@contextlib.contextmanager
def stdout_isolated():
    """Redirects the process's REAL stdout (file descriptor 1, not just the
    Python-level sys.stdout object) to /dev/null for the duration of the
    block. Necessary because mlx-audio prints plain (ANSI-colored) status
    lines straight to stdout even with verbose=False — confirmed directly:
    reassigning sys.stdout alone does NOT catch everything some libraries
    write, since certain output goes straight to the fd. Stdout here is our
    ONLY channel back to Go (one JSON object per line); anything else landing
    on it corrupts the protocol. Progress bars (tqdm) already default to
    stderr and are unaffected — this only isolates stdout.
    """
    sys.stdout.flush()
    saved_fd = os.dup(1)
    devnull_fd = os.open(os.devnull, os.O_WRONLY)
    os.dup2(devnull_fd, 1)
    os.close(devnull_fd)
    try:
        yield
    finally:
        sys.stdout.flush()
        os.dup2(saved_fd, 1)
        os.close(saved_fd)


def main():
    # Go waits for this exact line on stderr before sending any requests —
    # it's the signal that imports have finished and the process can accept
    # work, distinct from stdout (which carries only JSON responses).
    sys.stderr.write("WORKER_READY\n")
    sys.stderr.flush()

    # MLX's cache limit defaults to the memory limit (i.e. unbounded) — its
    # own docs say so plainly. That's fine for a short-lived script, but this
    # process stays warm for up to voiceWorkerIdleTimeout (15 real minutes,
    # see voice_worker.go) and can field many calls in that window: every
    # live-preview tick during a recording (every ~1.2s — see
    # LIVE_PREVIEW_INTERVAL_MS in useMicDictation.ts) on top of every real
    # WhatsApp voice note and mic dictation. On Apple Silicon's UNIFIED memory,
    # MLX's growing cache is the SAME physical RAM everything else on the
    # machine needs — left unmanaged, a long voice-heavy session can run the
    # whole system out of memory, not just this process. mx.clear_cache()
    # after every single request keeps this process's footprint bounded by
    # "one inference's worth," not "every inference this process has ever run."
    import mlx.core as mx

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            req = json.loads(line)
            with stdout_isolated():
                resp = HANDLERS[req["cmd"]](req)
        except Exception as e:  # noqa: BLE001 - reported to the caller, not swallowed
            resp = {"error": f"{type(e).__name__}: {e}"}
        finally:
            mx.clear_cache()
        sys.stdout.write(json.dumps(resp) + "\n")
        sys.stdout.flush()


if __name__ == "__main__":
    main()
